### ISSUE TITLE
Job is executed twice despite DisableConcurrentExecution

### ISSUE THREAD

**Issue by caveofjulian** [HangfireIO/Hangfire#1960]
I am using Hangfire.AspNetCore 1.7.17 and Hangfire.MySqlStorage 2.0.3 for software that is currently in production.

Now and then, we get a report of jobs being executed twice, despite the usage of the [DisableConcurrentExecution] attribute with a timeout of 30 seconds.

It seems that as soon as those 30 seconds have passed, another worker picks up that same job again.

The code is fairly straightforward:

```
public async Task ProcessPicking(HttpRequest incomingRequest)
{
    var filePath = await StoreStreamAsync(incomingRequest, TriggerTypes.Picking);
    var picking = await XmlHelper.DeserializeFileAsync<Picking>(filePath);
    // delay with 20 minutes so outbound-out gets the chance to be send first
    BackgroundJob.Schedule(() => StartPicking(picking), TimeSpan.FromMinutes(20));
}

[TriggerAlarming("[IMPORTANT] Failed to parse picking message to **** object.")]
[DisableConcurrentExecution(30)]
public void StartPicking(Picking picking)
{
    var orderlinePickModels = picking.ToSalesOrderlinePickQuantityRequests().ToList();
    var orderlineStatusModels = orderlinePickModels.ToSalesOrderlineStatusRequests().ToList();

    var isParsed = DateTime.TryParse(picking.Order.UnloadingDate, out var unloadingDate);

    for (var i = 0; i < orderlinePickModels.Count; i++)
    {
        // prevents bugs with usage of i in the background jobs
        var index = i;
        var id = BackgroundJob.Enqueue(() => SendSalesOrderlinePickQuantityRequest(orderlinePickModels[index], picking.EdiReference));
        BackgroundJob.ContinueJobWith(id, () => SendSalesOrderlineStatusRequest(
        orderlineStatusModels.First(x=>x.SalesOrderlineId== orderlinePickModels[index].OrderlineId), 
        picking.EdiReference, picking.Order.PrimaryReference, isParsed ? unloadingDate : DateTime.MinValue));
    }
}

[TriggerAlarming("[IMPORTANT] Failed to send order line pick quantity request to ****.")]
[AutomaticRetry(Attempts = 2)]
[DisableConcurrentExecution(30)]
public void SendSalesOrderlinePickQuantityRequest(SalesOrderlinePickQuantityRequest request, string ediReference)
{
    var audit = new AuditPostModel
    {
        Description = $"Finished job to send order line pick quantity request for item {request.Itemcode}, part of ediReference {ediReference}.",
        Object = request,
        Type = AuditTypes.SalesOrderlinePickQuantity
    };

    try
    {
        _logger.LogInformation($"Started job to send order line pick quantity request for item {request.Itemcode}.");
                
        var response = _service.SendSalesOrderLinePickQuantity(request).GetAwaiter().GetResult();
        audit.StatusCode = (int)response.StatusCode;
        if (!response.IsSuccessStatusCode) throw new TriggerRequestFailedException();
                
        audit.IsSuccessful = true;
        _logger.LogInformation("Successfully posted sales order line pick quantity request to ***** endpoint.");
    }
    finally
    {
        Audit(audit);
    }
}
```

It schedules the main task (StartPicking) that creates the objects required for the two subtasks:

Send picking details to customer
Send statusupdate to customer
The first job is duplicated. Perhaps the second job as well, but this is not important enough to care about as it just concerns a statusupdate. However, the first job causes the customer to think that more items have been picked than in reality.

I would assume that Hangfire updates the state of a job to e.g. in progress, and checks this state before starting a job. Is my time-out on the disabled concurrent execution too low? Is it possible in this scenario that the database connection to update the state takes about 30 seconds (to be fair, it is running on a slow server with ~8GB Ram, 6 vCores) due to which the second worker is already picking up the job again?

Or is this a Hangfire specific issue that must be tackled?

**Comment by AminSojoudi** [comment]
+1 here, we are using `Hangfire 1.7.13` and `Hangfire.AspNetCore 1.7.13` and  `Hangfire.Mongo 0.7.12`. 
Some jobs run twice with two server instances where the expected behavior is to run only once.
We do not use `DisableConcurrentExecution` attr. is it needed for this kind of behavior?
Is it something that related to `HangfireMongo` or `Hangfire`?

**Comment by caveofjulian** [comment]
> +1 here, we are using `Hangfire 1.7.13` and `Hangfire.AspNetCore 1.7.13` and `Hangfire.Mongo 0.7.12`. Some jobs run twice with two server instances where the expected behavior is to run only once. We do not use `DisableConcurrentExecution` attr. is it needed for this kind of behavior? Is it something that related to `HangfireMongo` or `Hangfire`?

To be honest, I have found quite some bugs now in Hangfire. I think the fact that people tell us to use DisableConcurrentExecution should have been an indication not to go with Hangfire. Unfortunately that choice is now too late. I implemented in-memory checks before execution of jobs to avoid it.

**Comment by Frazi1** [comment]
Hangfire is designed to run with `at-least-once` execution semantics. If it considers your job to be abandoned (for instance, because the processing server was shutdown and stopped sending heartbeats), then it will allow the same job to be picked by another worker.

This behavior highly depends on the storage you're using. With the latest version of `SqlServer` storage, everything should work like a charm out of the box.

Other storages may have issues. For example, `Postgres` storage has `InvisibilityTimeout` setting. If your job is taking longer than the timeout, then it will be constantly duplicated by another worker. Therefore, the solution is to increase the timeout.

I don't know much about `MySql` storage, though, but for me it seems really unstable. I recommend you to search its documentation for Invisibility Timeout and play around with that setting. If it does not help, it's better to switch to another storage.

**Comment by flmbray** [comment]
> This behavior highly depends on the storage you're using. With the latest version of `SqlServer` storage, everything should work like a charm out of the box.

It seems to me that different behavior based on the way Hangfire is storing the jobs is not a good design.  It certainly is giving me serious thoughts about dropping Hangfire completely.  

It's actually quite trivial to demonstrate the multiple-execution behavior - just set up a recurring job (even with Memory Storage) that executes every 1 minute, and the put a Thread.Sleep(70 * 1000) so that it takes 70 seconds.  A bit of extra logging easily shows that two of that thread will run at the same time, clearly _not_ what "DisableConcurrentExecution" seems to imply.

**Comment by resulkilic** [comment]
I solved this problem. I hope it works for you

See: https://stackoverflow.com/a/74334104/20430576

**Comment by resulkilic** [comment]
This was a problem I was having too. I solved this in my own project, I hope it works for you.

See: https://stackoverflow.com/a/74334104/20430576

**Comment by caveofjulian** [comment]
@resulkilic your code looks promising. I fortunately handed the project over to another programmer. He told me he forked this MySql library and did a fix in the source code itself. Im not sure if he pull requested, probably not.

**Comment by DaGrisa** [comment]
Same problem here, will this be fixed in the future or is it "as designed"?

**Comment by resulkilic** [comment]
The purpose of this is to prevent it from working again when there is a working job.

**Comment by DaGrisa** [comment]
Makes sense, but in our case the "waiting" job uses a worker, which leads to filling up all workers with "waiting" jobs and freezing the system.

**Comment by resulkilic** [comment]
No, that's exactly the point of this filter. Unnecessarily preventing the tail from swelling. I advise you to try. Trigger a task again while running to see the result and you will see it not added.

**Comment by DaGrisa** [comment]
> No, that's exactly the point of this filter. Unnecessarily preventing the tail from swelling. I advise you to try. Trigger a task again while running to see the result and you will see it not added.

Okay, I was talking about the problem in general. The filter does not work with our configuration.

**Comment by VmPeter** [comment]
flmbray:

As I understand, when the scheduled job triggers every minute, it triggers another job with another job id, so it's not the same job, so this scenario is not affected by DisableConcurrentExecution. I believe this is what Mutexes are for in their paid version https://docs.hangfire.io/en/latest/background-processing/throttling.html?highlight=disableconcurrentexecution#mutexes
But you could just as well use something like https://github.com/madelson/DistributedLock to make sure you are not running the same job type more than once even in a distributed system.

caveofjulian:

I'm unsure exactly how this double processing happened. Could it be that ProcessPicking was called twice? In that case DisableConcurrentExecution wouldn't help you, see above.

This seems to be connected to some of the common challenges when working with distributed systems which Hangfire suddenly becomes when having the DB as a middle-man between executions on multiple servers.

The notions of an "at least once", "at most once" and "only once"-delivery etc. AFAIK is extremely difficult, if not impossible to get "only once" guaranteed message delivery when communicating over an unreliable link (see https://en.wikipedia.org/wiki/Two_Generals%27_Problem or https://www.youtube.com/watch?v=IP-rGJKSZ3s), even RabbitMQ doesn't claim this. https://www.rabbitmq.com/reliability.html

This is why most resort to their methods being idempotent and these things working on a best-effort basis.

Maybe you can try another approach here, to try to design around this happening.

Looks like ProcessPicking is called via a http request with a batch of sales orders to be processed? And I'm assuming this should end up in an e-mail to multiple customers, and you don't want multiple e-mails per customer.

If the file-size is big, you'll have a huge Picking class, so when you enqueue StartPicking, Hangfire will json serialize this and put it into "Arguments" columns for this job. I suspect this could cause issues if 1 row has a size of multiple megabytes.

Instead try to enqueue the n sales orders directly in ProcessPicking. Then each job will be quick to execute and will send exactly 1 e-mail. Thus even if one job should be re-tried, it would only be 1 e-mail, vs. in your case should StartPicking be retried, it will resend many.

You still have potential issues, like multiple Hangfire servers picking up the same job, if this was in fact was happened. If this is the case, then a distributed lock on the functionname + customer number might work, but I would try the above first before doing this.

You would still have the possibility that enqueuing half-way through fails in ProcessPicking, then caller would get non-ok and may call again with the same sales orders, so half would be enqueued twice. So you also have to design yourself out of that.

One thing you could do was to only take in 1 sales order at a time in the request. The caller would then get OK only on those that are enqueued, and would retry those that are not. If that is not feasible, you would have to return a list of ids that are successfully processed, and those that are not. Again it's up to the caller to retry only those that fail.

**Comment by peheje** [comment]
flmbray:

As I understand, when the scheduled job triggers every minute, it triggers another job with another job id, so it's not the same job, so this scenario is not affected by DisableConcurrentExecution. I believe this is what Mutexes are for in their paid version https://docs.hangfire.io/en/latest/background-processing/throttling.html?highlight=disableconcurrentexecution#mutexes
But you could just as well use something like https://github.com/madelson/DistributedLock to make sure you are not running the same job type more than once.

caveofjulian:

I'm unsure exactly how this double processing happened. Could it be that ProcessPicking was called twice? In that case DisableConcurrentExecution wouldn't help you, see above.

You also have issues in the loop if Enqueue fails half-way through, Hangfire would retry StartPicking from the beginning, sending multiple e-mails. In short I don't think the code-design is ideal.

This seems to be connected to some of the common challenges when working with distributed systems which Hangfire suddenly becomes when having the DB as a middle-man between executions on multiple servers.

The notions of an "at least once", "at most once" and "only once"-delivery etc. AFAIK is extremely difficult, if not impossible to get "only once" guaranteed message delivery when communicating over an unreliable link (see https://en.wikipedia.org/wiki/Two_Generals%27_Problem or https://www.youtube.com/watch?v=IP-rGJKSZ3s), even RabbitMQ doesn't claim this. https://www.rabbitmq.com/reliability.html

This is why most resort to their methods being idempotent and these things working on a best-effort basis.

Maybe you can try another approach here, to try to design around this happening.

Looks like ProcessPicking is called via a http request with a batch of sales orders to be processed? And I'm assuming this should end up in an e-mail to multiple customers, and you don't want multiple e-mails per customer.

If the file-size is big, you'll have a huge Picking class, so when you enqueue StartPicking, Hangfire will json serialize this and put it into "Arguments" columns for this job. I suspect this could cause issues if 1 row has a size of multiple megabytes.

Instead enqueue the n sales orders directly in ProcessPicking. Then each job will be quick to execute and will send exactly 1 e-mail. Thus even if one job should be re-tried, it would only be 1 e-mail, vs. in your case should StartPicking be retried, it will resend many.

You still have potential issues, like multiple Hangfire servers picking up the same job, if this was in fact was happened. If this is the case, then a distributed lock on the functionname + customer number might work, but I would try the above first before doing this.

You would still have the possibility that enqueuing half-way through fails in ProcessPicking, then caller would get non-ok and may call again with the same sales orders, so half would be enqueued twice. So you also have to design yourself out of that.

One thing you could do was to only take in 1 sales order at a time in the request. The caller would then get OK only on those that are enqueued, and would retry those that are not. If that is not feasible, you would have to return a list of ids that are successfully processed, and those that are not. Again it's up to the caller to retry only those that fail.

**Comment by Gerappa92** [comment]
> > This behavior highly depends on the storage you're using. With the latest version of `SqlServer` storage, everything should work like a charm out of the box.
> 
> It seems to me that different behavior based on the way Hangfire is storing the jobs is not a good design. It certainly is giving me serious thoughts about dropping Hangfire completely.
> 
> It's actually quite trivial to demonstrate the multiple-execution behavior - just set up a recurring job (even with Memory Storage) that executes every 1 minute, and then put a Thread.Sleep(70 * 1000) so that it takes 70 seconds. A bit of extra logging easily shows that two of that threads will run at the same time, clearly _not_ what "DisableConcurrentExecution" seems to imply.

I have run into this scenario. But in my case, it was placing an attribute in the wrong place. I was creating the recurring job in the following way:

`RecurringJob.AddOrUpdate<IWebApiJobExecutor>(job.Id.ToString(), svc => svc.Call(webApiCallback), job.Cron);`

And the attribute works for me only when I place it above my interface:

```
[DisableConcurrentExecution(60)]
internal interface IWebApiJobExecutor
```

Adding this attribute to the implementation or method didn't work.

[Link to stachoverflow](https://stackoverflow.com/a/74656892/6312282)

**Comment by billpeace** [comment]
Please try using the [SkipWhenPreviousJobIsRunningAttribute]( https://gist.github.com/odinserj/a6ad7ba6686076c9b9b2e03fcf6bf74e ) job filter from the referenced gist

**Comment by simeyla** [comment]
Consider also [`LatencyTimeout`](https://github.com/HangfireIO/Hangfire/blob/master/src/Hangfire.Core/LatencyTimeoutAttribute.cs) which will delete the job after the timeout has elapsed.
This is good for cleanup jobs where you want the job to run every 5 minute (eg. `CleanupPayments`) but you don't need or want it running ten times all at once if the server was down for an hour.

For a job that runs every ten minutes I might set this to 1 minute or even shorter. Depending upon your needs it could be anything up to the full ten minutes. And of course you can also apply `DisableConcurrentExecution` just to be sure.

You wouldn't of course want to use this for jobs that have to complete - such as `ProcessOrder`.

This has confused me many times. Would be nice if we could define named profiles for the type of job, which would incorporate all such rules.

**Comment by wswind** [comment]
try this solution

```cs
// Source - https://stackoverflow.com/a/79956874
// Posted by ws_
// Retrieved 2026-06-12, License - CC BY-SA 4.0

public class SkipConcurrentExecution : JobFilterAttribute, IServerFilter
{
    const int MaxProcessingCount = 100;
    public void OnPerforming(PerformingContext filterContext)
    {
        var jobs = JobStorage.Current.GetMonitoringApi().ProcessingJobs(0, MaxProcessingCount);
        if (jobs.Any(x => x.Key != filterContext.BackgroundJob.Id &&
                          x.Value.Job.Type == filterContext.BackgroundJob.Job.Type &&
                          x.Value.Job.Method.Name == filterContext.BackgroundJob.Job.Method.Name &&
                          x.Value.Job.Args.Count == filterContext.BackgroundJob.Job.Args.Count &&
                          string.Join(",", x.Value.Job.Args) == string.Join(",", filterContext.BackgroundJob.Job.Args)))
        {
            filterContext.Canceled = true;
        }
    }

    public void OnPerformed(PerformedContext filterContext) { }
}

```
